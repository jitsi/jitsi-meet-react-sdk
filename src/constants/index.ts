export const DEFAULT_DOMAIN = 'meet.jit.si';

export const DEFAULT_OPTIONS = {
  roomName: 'DefaultRoomNameExample',
  width: '100%',
  height: '700px'
};

export const JAAS_DOMAIN = '8x8.vc';
