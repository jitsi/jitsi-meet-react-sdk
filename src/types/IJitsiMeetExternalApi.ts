import EventEmitter from 'events';

export default interface IJitsiMeetExternalApi extends EventEmitter {

    /**
     * Adds an event listener to the Jitsi Meet External API.
     * @deprecated backward compatibility only, preferably use `on` instead
     */
    addEventListener: (event: string, fn: () => void) => void;

    /**
     * Adds multiple event listeners to the Jitsi Meet External API.
     * @deprecated backward compatibility only
     */
    addEventListeners: (listeners: object) => void;

    /**
     * Captures the screenshot of the large video.
     * Resolves with a base64 encoded image data of the screenshot
     * if large video is detected, throws an error otherwise.
     */
    captureLargeVideoScreenshot: () => Promise<string>;

    /**
     * Removes the listeners and removes the Jitsi Meet frame.
     *
     */
    dispose: () => void;

    /**
     * Executes command.
     */
    executeCommand: (name: string, ...args: any[] | undefined[]) => void;

    /**
     * Executes commands.
     * The keys of the object parameter are the commands that will be executed and the
     * values are the arguments for the command.
     */
    executeCommands: (commands: { [command: string]: any[] | undefined[] }) => void;

    /**
     * Returns Promise that resolves with a list of available devices.
     */
    getAvailableDevices: () => Promise<object[]>;

    /**
     * Returns the avatar URL of a participant.
     */
    getAvatarURL: (participantId: string) => string;

    /**
     * Gets a list of the currently sharing participant ids.
     */
    getContentSharingParticipants: () => Promise<string[]>;

    /**
     * Returns Promise that resolves with current selected devices.
     */
    getCurrentDevices: () => Promise<object[]>;

    /**
     * Returns any custom avatars backgrounds.
     */
    getCustomAvatarBackgrounds: () => Promise<string[]>;

    /**
     * Gets the deployment info.
     */
    getDeploymentInfo: () => Promise<object>;

    /**
     * Returns the display name of a participant.
     */
    getDisplayName: (participantId: string) => string;

    /**
     * Returns the email of a participant.
     */
    getEmail: (participantId: string) => string;

    /**
     * Returns the IFrame that loads the Jitsi Meet External API.
     */
    getIFrame: () => HTMLElement;

    /**
     * Returns the current livestream url.
     * Resolves with the current livestream URL if exists, with
     * undefined if not, and rejects on failure.
     */
    getLivestreamUrl: () => Promise<string | undefined>

    /**
     * Returns the number of participants in the conference.
     * The local participant is included.
     */
    getNumberOfParticipants: () => number;

    /**
     * Returns the conference participants information as an array containing participants
     * information like participant id, display name, avatar URL and email.
     */
    getParticipantsInfo: () => object[];

    /**
     * Returns the current video quality setting.
     */
    getVideoQuality: () => number;

    /**
     * Invite people to the call.
     */
    invite: (invitees: object[]) => Promise<object[]>;

    /**
     * Check if the audio is available.
     */
    isAudioAvailable: () => Promise<boolean>;

    /**
     * Returns the audio mute status.
     */
    isAudioMuted: () => Promise<boolean>;

    /**
     * Returns whether the device change is available.
     */
    isDeviceChangeAvailable: () => Promise<boolean>;

    /**
     * Returns whether the device list is available.
     */
    isDeviceListAvailable: () => Promise<boolean>;

    /**
     * Returns the moderation on status on the given mediaType.
     */
    isModerationOn: () => Promise<boolean>;

    /**
     * Returns whether multiple audio input is supported.
     */
    isMultipleAudioInputSupported: () => Promise<boolean>;

    /**
     * Returns force muted status of the given participant id for the given media type.
     */
    isParticipantForceMuted: (participantId: string, mediaType: string) => Promise<boolean>;

    /**
     * Returns screen sharing status.
     */
    isSharingScreen: () => Promise<boolean>;

    /**
     * Check if the video is available.
     */
    isVideoAvailable: () => Promise<boolean>;

    /**
     * Returns the audio mute status.
     */
    isVideoMuted: () => Promise<boolean>;

    /**
     * Pins a participant's video on to the stage view.
     */
    pinParticipant: (participantId: string) => void;

    /**
     * Removes the given event listener.
     */
    removeEventListener: (event: string) => void;

    /**
     * Removes event listeners for the given list.
     */
    removeEventListeners: (eventList: string[]) => void;

    /**
     * Resizes the large video container as per the dimensions provided.
     */
    resizeLargeVideo: (width: number, height: number) => void;

    /**
     * Passes an event along to the local conference participant to establish
     * or update a direct peer connection. This is currently used for developing
     * wireless screensharing with room integration and it is advised against to
     * use as its api may change.
     */
    sendProxyConnectionEvent: (event: {

        /**
         * The payload of the event.
         */
        data: object,

        /**
         * The JID of the sender of the event.
         * Needed when a reply is to be sent regarding the event.
         */
        form: string
    }) => void;

    /**
     * Sets the audio input device to the one with the label or id that is passed.
     */
    setAudioInputDevice: (label: string, deviceId: string) => Promise<void>;

    /**
     * Sets the audio output device to the one with the label or id that is passed.
     */
    setAudioOutputDevice: (label: string, deviceId: string) => Promise<void>;

    /**
     * Displays the given participant on the large video. If no participant id is specified,
     * dominant and pinned speakers will be taken into consideration while selecting the
     * the large video participant.
     */
    setLargeVideoParticipant: (participantId: string) => void;

    /**
     * Sets the video input device to the one with the label or id that is passed.
     */
    setVideoInputDevice: (label: string, deviceId: string) => Promise<void>;

    /**
     * Starts a file recording or streaming session depending on the passed on params.
     * * For RTMP streams, `rtmpStreamKey` must be passed on. `rtmpBroadcastID` is optional.
     * * For youtube streams, `youtubeStreamKey` must be passed on. `youtubeBroadcastID` is optional.
     * * For dropbox recording, recording `mode` should be `file` and a dropbox oauth2 token must be provided.
     * * For file recording, recording `mode` should be `file` and optionally `shouldShare` could be passed on.
     *
     * No other params should be passed.
     */
    startRecording: (options: {

        /**
         * The recording mode.
         */
        mode: 'file' | 'stream',

        /**
         * The dropbox oauth2 token.
         */
        dropboxToken: string,

        /**
         * Whether the recording should be shared with the participants or not.
         * Only applies to certain Jitsi Meet deploys.
         */
        shouldShare: boolean,

        /**
         * The RTMP stream key.
         */
        rtmpStreamKey: string,

        /**
         * The RTMP broacast ID.
         */
        rtmpBroadcastID: string,

        /**
         * The youtube stream key.
         */
        youtubeStreamKey: string,

        /**
         * The youtube broacast ID.
         */
        youtubeBroadcastID: string
    }) => void;

    /**
     * Stops a recording or streaming session that is in progress.
     */
    stopRecording: (mode: 'file' | 'stream') => void;
}
