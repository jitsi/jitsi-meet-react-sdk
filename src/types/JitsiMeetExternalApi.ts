import { IJitsiMeetExternalApi } from '.';

export type JitsiMeetExternalApi = {
    new (

        /**
         * The domain used to build the conference URL.
         */
        domain: string,

        /**
         * The optional arguments for the IFrame.
         */
        options: {

            /**
             * The name of the room to join.
             */
            roomName?: string,

            /**
             * The created IFrame width.
             */
            width?: string | number,

            /**
             * The height for the created IFrame.
             */
            height?: string | number,

            /**
             * The JS object with overrides for options defined in the config.js file
             * https://github.com/jitsi/jitsi-meet/blob/master/config.js.
             */
            configOverwrite?: object,

            /**
             * The JS object with overrides for options defined in the interface_config.js file
             * https://github.com/jitsi/jitsi-meet/blob/master/interface_config.js.
             */
            interfaceConfigOverwrite?: object;

            /**
             * The JWT token.
             */
            jwt?: string,

            /**
             * The IFrame onload event handler.
             */
            onload?: () => void,

            /**
             * Object arrays that contain information about participants invited to a call.
             */
            invitees?: [],

            /**
             * Information map about the devices used in a call.
             */
            devices?: {

                /**
                 * The label of the device used for audio input.
                 */
                audioInput: string,

                /**
                 * The label of the device  used for audio output.
                 */
                audioOutput: string,

                /**
                 * The label of the device used for video input.
                 */
                videoInput: string
            },

            /**
             * The JS object that contains information about the participant starting the meeting.
             */
            userInfo?: {

                /**
                 * The participant display name.
                 */
                displayName: string,

                /**
                 * The participant email address.
                 */
                email: string
            },

            /**
             * The `stage.8x8.vc` or `8x8.vc` release version to test.
             * Expects the following format: `release-1234`.
             */
            release?: string,

            /**
             * The HTML DOM Element where the IFrame is added as a child.
             */
            parentNode?: HTMLDivElement | null
        }
    ): IJitsiMeetExternalApi
};
