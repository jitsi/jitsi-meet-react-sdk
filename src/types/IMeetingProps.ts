import React from 'react';

import IJitsiMeetExternalApi from './IJitsiMeetExternalApi';

/**
 * The the base props for React {@code Components} that load the Jitsi Meet IFrame.
 */
export default interface IMeetingProps {

    /**
     * The name of the room to join.
     */
    roomName: string;

    /**
     * The JS object with overrides for options defined in the config.js file
     * https://github.com/jitsi/jitsi-meet/blob/master/config.js.
     */
    configOverwrite?: object;

    /**
     * The JS object with overrides for options defined in the interface_config.js file
     * https://github.com/jitsi/jitsi-meet/blob/master/interface_config.js.
     */
    interfaceConfigOverwrite?: object;

    /**
     * The JWT token.
     */
    jwt?: string;

    /**
     * Object arrays that contain information about participants invited to a call.
     */
    invitees?: [];

    /**
     * Information map about the devices used in a call.
     */
    devices?: {

        /**
         * The label of the device used for audio input.
         */
        audioInput: string;

        /**
         * The label of the device  used for audio output.
         */
        audioOutput: string;

        /**
         * The label of the device used for video input.
         */
        videoInput: string;
    };

    /**
     * The JS object that contains information about the participant starting the meeting.
     */
    userInfo?: {

        /**
         * The participant display name.
         */
        displayName: string;

        /**
         * The participant email address.
         */
        email: string;
    };

    /**
     * The `stage.8x8.vc` or `8x8.vc` release version to test.
     * Expects the following format: `release-1234`.
     */
    release?: string;

    /**
     * The custom spinner to be displayed while the iframe is loading.
     */
    spinner?: React.ElementType;

    /**
     * The external API reference for events and commands.
     */
    onApiReady?: (api: IJitsiMeetExternalApi) => void;

    /**
     * The callback for when the meeting is ready to be closed.
     */
    onReadyToClose?: () => void;

    /**
     * The parent node used by the IFrame.
     */
    getIFrameRef?: (parentNode: HTMLDivElement) => void;
}
