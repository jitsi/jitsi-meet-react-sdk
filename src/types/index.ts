import { JitsiMeetExternalApi } from './JitsiMeetExternalApi';
declare global {
    interface Window {
        JitsiMeetExternalAPI: JitsiMeetExternalApi;
    }
}

export { JitsiMeetExternalApi } from './JitsiMeetExternalApi';
export { default as IJitsiMeetingProps } from './IJitsiMeetingProps';
export { default as IJaaSMeetingProps } from './IJaaSMeetingProps';
export { default as IJitsiMeetExternalApi } from './IJitsiMeetExternalApi';

