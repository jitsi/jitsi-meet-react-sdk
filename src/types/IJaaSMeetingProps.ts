/* eslint-disable-next-line */
import { JaaSMeeting } from '..';
import IMeetingProps from './IMeetingProps';

/**
 * The type of the React {@code Component} props of {@link JaaSMeeting}.
 */
export default interface IJaaSMeetingProps extends IMeetingProps {

    /**
     * The App ID that provides an isolated context and prefixes the room name.
     */
    appId: string;

    /**
     * Whether to use the stage environment or not.
     */
    useStaging?: boolean;
}
