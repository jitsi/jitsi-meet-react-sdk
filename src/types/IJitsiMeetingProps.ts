/* eslint-disable-next-line */
import { JitsiMeeting } from '..';
import IMeetingProps from './IMeetingProps';

/**
 * The type of the React {@code Component} props of {@link JitsiMeeting}.
 */
export default interface IJitsiMeetingProps extends IMeetingProps {

    /**
     * The domain used to build the conference URL.
     */
    domain?: string;
}
