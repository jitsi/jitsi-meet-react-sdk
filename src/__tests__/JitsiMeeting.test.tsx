import * as React from 'react'
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { JitsiMeeting } from '../index';
import { DEFAULT_DOMAIN, DEFAULT_OPTIONS } from '../constants';
import { IJitsiMeetExternalApi } from '../types';

describe('JitsiMeeting component', () => {
  it('should render correctly', () => {
    const props = {
      ...DEFAULT_OPTIONS,
      domain: DEFAULT_DOMAIN,
      onApiReady: (externalApi: IJitsiMeetExternalApi) => { console.log(externalApi); }
    };

    const wrapper = shallow(<JitsiMeeting {...props} />);
    const snapshot = toJson(wrapper);
    // Overwrite non-deterministic fields
    snapshot.children[0].props.id = 'test';
    snapshot.children[0].props.key = 'test';
    expect(snapshot).toMatchSnapshot();
  });

  it('should render iframe', () => {
    const props = {
      ...DEFAULT_OPTIONS,
      domain: DEFAULT_DOMAIN,
      onApiReady: (externalApi: IJitsiMeetExternalApi) => { console.log(externalApi); }
    };
    const wrapper = shallow(<JitsiMeeting {...props} />);
    const iframe = wrapper.find('iframe');
    expect(iframe).toBeTruthy();
  });
});