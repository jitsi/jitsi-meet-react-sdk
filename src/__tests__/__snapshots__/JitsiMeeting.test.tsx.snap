// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JitsiMeeting component should render correctly 1`] = `
<Fragment>
  <div
    id="test"
    key="test"
  />
</Fragment>
`;
